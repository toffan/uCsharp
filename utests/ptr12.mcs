typedef struct {int x; int y;} Point;
typedef struct {Point* c; int r;} Cercle;
typedef struct {Point a; Point b;} Segment;

int* malloc(int size) {
    int* addr;
    asm {
        LOAD %size
        SUBR Malloc
        STORE %addr
    }
    return addr;
}

int main() {
    Point* pt1 = (Point*) malloc(2);
    Cercle* c1 = (Cercle*) malloc(2);
    pt1->y = 8;
    (*pt1).x = 4;
    c1->c = pt1;
    Segment* s1 = (Segment*) malloc(4);
    s1->a = *pt1;
    s1->b.x = 1;
    s1->b.y = s1->a.y;

    return 0;
}
