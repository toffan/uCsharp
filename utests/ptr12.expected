JUMP __malloc
_malloc:
    PUSH 1
	LOAD (1) -1[LB]
	SUBR Malloc
	STORE (1) 3[LB]
    LOAD (1) 3[LB]
    RETURN (1) 1
    RETURN (0) 1
__malloc:

JUMP __main
_main:
    LOADL 2
    CALL (SB) _malloc
    LOADL 2
    CALL (SB) _malloc

    LOADL 8
    LOAD (1) -1[ST]
    LOAD (1) 3[LB]
    LOADL 1
    SUBR IAdd
    STOREI (1)
    POP (0) 1

    LOADL 4
    LOAD (1) -1[ST]
    LOAD (1) 3[LB]
    LOADL 0
    SUBR IAdd
    STOREI (1)
    POP (0) 1

    LOAD (1) 3[LB]
    LOAD (1) -1[ST]
    LOAD (1) 4[LB]
    LOADL 0
    SUBR IAdd
    STOREI (1)
    POP (0) 1

    LOADL 4
    CALL (SB) _malloc

    LOAD (1) 3[LB]
    LOADI (2)
    LOAD (2) -2[ST]
    LOAD (1) 5[LB]
    LOADL 0
    SUBR IAdd
    STOREI (2)
    POP (0) 2

    LOADL 1
    LOAD (1) -1[ST]
    LOAD (1) 5[LB]
    LOADL 2
    SUBR IAdd
    LOADL 0
    SUBR IAdd
    STOREI (1)
    POP (0) 1

    LOAD (1) 5[LB]
    LOADL 0
    SUBR IAdd
    LOADL 1
    SUBR IAdd
    LOADI (1)
    LOAD (1) -1[ST]
    LOAD (1) 5[LB]
    LOADL 2
    SUBR IAdd
    LOADL 1
    SUBR IAdd
    STOREI (1)
    POP (0) 1

    LOADL 0
    RETURN (1) 0
    RETURN (0) 0
__main:
    CALL (SB) _main
    HALT
